tensorboard/plugins/graph/graph_util_test.py

```diff
@@ -0,0 +1,488 @@
+              type: DT_HALF
+            }
+          }
+          node_def {
+            name: "add"
+            op: "Add"
+            input: "x"
+            input: "y"
+          }
+        }
+      }
+    ''', graph_def_a)
+
+    graph_def_b = GraphDef()
+    text_format.Merge('''
+      library {
+        function {
+          signature {
+            name: "foo"
+            input_arg {
+              name: "x"
+              type: DT_HALF
+            }
+            output_arg {
+              name: "identity"
+              type: DT_HALF
+            }
+          }
+          node_def {
+            name: "add"
+            op: "Add"
+            input: "x"
+            input: "y"
+          }
+        }
+        function {
+          signature {
+            name: "foo_1"
+            input_arg {
+              name: "x"
+              type: DT_HALF
+            }
+            output_arg {
+              name: "identity"
+              type: DT_HALF
+            }
+          }
+          node_def {
+            name: "add"
+            op: "Add"
+            input: "x"
+            input: "y"
+          }
+        }
+      }
+    ''', graph_def_b)
+
+    self.assertProtoEquals(
+        expected_proto,
+        graph_util.combine_graph_defs(graph_def_a, graph_def_b))
+
+  def test_combine_graph_defs_function_collison(self):
+    graph_def_a = GraphDef()
+    text_format.Merge('''
+      library {
+        function {
+          signature {
+            name: "foo"
+            input_arg {
+              name: "x"
+              type: DT_HALF
+            }
+            output_arg {
+              name: "identity"
+              type: DT_HALF
+            }
+          }
+          node_def {
+            name: "add"
+            op: "Add"
+            input: "x"
+            input: "y"
+          }
+        }
+      }
+    ''', graph_def_a)
+
+    graph_def_b = GraphDef()
+    text_format.Merge('''
+      library {
+        function {
+          signature {
+            name: "foo"
+            input_arg {
+              name: "x"
+              type: DT_HALF
+            }
+            output_arg {
+              name: "identity"
+              type: DT_HALF
+            }
+          }
+          node_def {
+            name: "div"
+            op: "Div"
+            input: "x"
+            input: "y"
+          }
+        }
+        function {
+          signature {
+            name: "foo_1"
+            input_arg {
+              name: "x"
+              type: DT_HALF
+            }
+            output_arg {
+              name: "identity"
+              type: DT_HALF
+            }
+          }
+          node_def {
+            name: "add"
+            op: "Add"
+            input: "x"
+            input: "y"
+          }
+        }
+      }
+    ''', graph_def_b)
+
+    with six.assertRaisesRegex(
+        self,
+        ValueError,
+        r'Cannot combine GraphDefs because functions share a name but are different. foo'):
```

Comment:
I don’t understand how this test is passing: the string “functions share
a name” doesn’t appear anywhere in the code under test. What’s going on?

Reply:
I had a spec with the same name and second one ran but not this one.

Reply:
Ugh, *that* particular nasty corner of Python. Good catch; thanks.

We should make that a lint error. It looks like we don’t have any such
violations in HEAD:

```
$ git ls-files '*_test.py' "$filename" | while read filename; do sed -n 's/.*\(def test[a-zA-Z0-9_]*\)(.*/\1/p; /\bclass\b/p' "$filename" | awk -v "filename=${filename}" '$1 == "class" { for (k in c) c[k] = 0; next; } $1 == "def" { if (c[$2]++) printf "%s: %s\n", filename, $2; next } { printf "unknown token: %s\n", $1 > "/dev/stderr"; exit 2; }'; done | diff -u /dev/null -
```